game-core/src/main/java/games/strategy/engine/framework/ServerGame.java

```diff
@@ -349,27 +351,20 @@ public void stopGame() {
     gameData.getGameLoader().shutDown();
   }
 
-  private void autoSave(final String fileName) {
-    final File autoSaveDir = new File(ClientSetting.SAVE_GAMES_FOLDER_PATH.value(), "autoSave");
-    // Above creates an autoSave subdirectorty pathname below the game games path
-    if (!autoSaveDir.exists()) {
-      autoSaveDir.mkdirs();
-    }
-    saveGame(new File(autoSaveDir, fileName));
-  }
-
-  private void autoSaveBefore(final IDelegate currentDelegate) {
-    autoSave(getAutoSaveBeforeFileName(currentDelegate.getName()));
-  }
-
-  private static String getAutoSaveBeforeFileName(final String stepName) {
-    final String baseFileName = "autosaveBefore" + stepName.substring(0, 1).toUpperCase() + stepName.substring(1);
-    return GameDataFileUtils.addExtension(baseFileName);
+  private void autoSaveBefore(final IDelegate delegate) {
+    saveGame(SaveGameFileChooser.getBeforeStepAutoSaveFile(delegate.getName()));
   }
 
   @Override
   public void saveGame(final File file) {
-    try (FileOutputStream fout = new FileOutputStream(file)) {
+    checkNotNull(file);
+
+    final File parentDir = file.getParentFile();
+    if (!parentDir.exists() && !parentDir.mkdirs()) {
+      log.severe("Failed to create save game directory (or one of its ancestors): " + parentDir.getAbsolutePath());
```

Comment:
Possibly.  The original code didn't even check the return value of `mkdirs()`.  I thought I was being a good citizen by logging the failure, but otherwise letting the true failure originate from the attempt to open the file on line 367.

Note also there's a TOCTOU between the `exists()` call and the `mkdirs()` call becuse the directory could be deleted between the two calls.  Likewise, there's a TOCTOU between `mkdirs()` and line 367 for the same reason, so that opening the file would still fail on that line.

Maybe that implies that the log level should be reduced to either `WARNING` or `INFO` (or even removed outright like the original code) and once again let the log on line 370 drive the observable failure.

> Alternatively throwing an exception instead would be more consistent with the behaviour of those methods when a securitymanager is active.

Agreed it makes more sense to let the caller decide how to handle this, but that would change the API, which seems to be out-of-scope for this PR.